File: dev.py
import glob
import os
import subprocess
import threading
import time
from http.server import ThreadingHTTPServer, SimpleHTTPRequestHandler

build_condition = threading.Condition()
build_version = 0

RELOAD_SCRIPT = b'<script>new EventSource("/~reload").onmessage=()=>location.reload()</script>'


class Handler(SimpleHTTPRequestHandler):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, directory="public", **kwargs)

    def do_GET(self):
        if self.path == "/~reload":
            self._sse()
            return
        path = self.translate_path(self.path)
        if os.path.isfile(path) and path.endswith(".html"):
            with open(path, "rb") as f:
                body = f.read()
            body = body.replace(b"</body>", RELOAD_SCRIPT + b"</body>", 1)
            self.send_response(200)
            self.send_header("Content-Type", "text/html; charset=utf-8")
            self.send_header("Content-Length", str(len(body)))
            self.end_headers()
            self.wfile.write(body)
            return
        super().do_GET()

    def _sse(self):
        self.send_response(200)
        self.send_header("Content-Type", "text/event-stream")
        self.send_header("Cache-Control", "no-cache")
        self.end_headers()
        local_version = build_version
        try:
            while True:
                with build_condition:
                    build_condition.wait_for(lambda: build_version > local_version)
                    local_version = build_version
                self.wfile.write(b"data: reload\n\n")
                self.wfile.flush()
        except Exception:
            pass

    def log_message(self, format, *args):
        pass


def mtimes():
    paths = (
        glob.glob("src/**", recursive=True)
        + ["gen.rb"]
        + glob.glob("view/**", recursive=True)
    )
    return {p: os.path.getmtime(p) for p in paths if os.path.isfile(p)}


def watch():
    global build_version
    prev = mtimes()
    while True:
        time.sleep(1)
        curr = mtimes()
        if curr != prev:
            subprocess.run(["make", "build"])
            with build_condition:
                build_version += 1
                build_condition.notify_all()
            prev = mtimes()
        else:
            prev = curr


if __name__ == "__main__":
    threading.Thread(target=watch, daemon=True).start()
    server = ThreadingHTTPServer(("", 8000), Handler)
    print("Serving at http://localhost:8000")
    server.serve_forever()

File: dev.rb
require 'webrick'

require_relative 'gen'

RELOAD_SCRIPT = '<script>new EventSource("/system/reload").onmessage=()=>location.reload()</script>'

# Fans builds out to the live-reload connections, one queue each.
class BuildSignal
  def initialize
    @mutex = Mutex.new
    @queues = []
  end

  # Returns a queue yielding one value per build, then nil once stopped.
  def subscribe
    Thread::Queue.new.tap { |queue| @mutex.synchronize { @queues << queue } }
  end

  def unsubscribe(queue)
    @mutex.synchronize { @queues.delete(queue) }
  end

  def bump
    @mutex.synchronize { @queues.each { |queue| queue.push(true) } }
  end

  # Releases the readers so that they don't hold up shutdown.
  def stop
    @mutex.synchronize { @queues.each(&:close) }
  end
end

# Serves public/, splicing the reload listener into every HTML page.
class ReloadingFileHandler < WEBrick::HTTPServlet::FileHandler
  def do_GET(req, res)
    super
    return unless res['content-type'].to_s.start_with?('text/html')

    body = res.body.respond_to?(:read) ? res.body.read : res.body.to_s
    res.body = body.sub('</body>', RELOAD_SCRIPT + '</body>')
    res.content_length = res.body.bytesize
  end
end

# Holds the connection open and sends an event after each build.
class ReloadServlet < WEBrick::HTTPServlet::AbstractServlet
  def initialize(server, signal)
    super(server)
    @signal = signal
  end

  def do_GET(_req, res)
    res['content-type'] = 'text/event-stream'
    res['cache-control'] = 'no-cache'
    res.chunked = true
    queue = @signal.subscribe
    res.body = proc do |out|
      out.write("data: reload\n\n") while queue.pop
    ensure
      @signal.unsubscribe(queue)
    end
  end
end

def mtimes
  paths = Dir.glob('src/**/*') + ['gen.rb'] + Dir.glob('view/**/*')
  paths.select { |path| File.file?(path) }.to_h { |path| [path, File.mtime(path)] }
end

def build
  Gen.build
rescue StandardError => e
  STDERR.puts("failed to build: #{e}")
end

def watch(signal)
  prev = mtimes
  loop do
    sleep 1
    next if (current = mtimes) == prev

    # The generator is loaded into this process, so its own edits need a restart.
    exec(RbConfig.ruby, $0) if current['gen.rb'] != prev['gen.rb']

    build
    signal.bump
    prev = mtimes
  end
end

if __FILE__ == $0
  build

  signal = BuildSignal.new
  Thread.new { watch(signal) }

  server = WEBrick::HTTPServer.new(
    Port: 8000,
    AccessLog: [],
    Logger: WEBrick::Log.new($stderr, WEBrick::Log::WARN)
  )
  server.mount('/', ReloadingFileHandler, 'public')
  server.mount('/system/reload', ReloadServlet, signal)

  # A queue push is one of the few things a trap handler may do; taking a lock
  # is not, and shutting down needs one.
  interrupts = Thread::Queue.new
  trap('INT') { interrupts.push(true) }
  Thread.new do
    interrupts.pop
    signal.stop # shutdown joins the connections, so let them finish first
    server.shutdown
  end

  puts 'Serving at http://localhost:8000'
  server.start
end
